saw-core/prelude/Prelude.sawcore

```diff
@@ -1389,6 +1474,287 @@ primitive mkDouble : Integer -> Integer -> Float;
+                          (atBVVec n len a v i pf));
+
+-- Take the first i elements of a BVVec, assuming i <= len
+takeBVVec : (n : Nat) -> (len : bitvector n) -> (a : sort 0) ->
+            (i : bitvector n) -> is_bvule n i len ->
+            BVVec n len a -> BVVec n i a;
+takeBVVec n len a i pf v =
+  genBVVec n i a
+    (\ (j:bitvector n) (pf2:is_bvult n j i) ->
+       atBVVec n len a v j (trans_bvult_bvule n j i len pf2 pf));
+
+-- Drop the first i elements of a BVVec, assuming i <= len
+dropBVVec : (n : Nat) -> (len : bitvector n) -> (a : sort 0) ->
+            (i : bitvector n) -> is_bvule n i len ->
+            BVVec n len a -> BVVec n (bvSub n len i) a;
+dropBVVec n len a i pf v =
+  genBVVec n (bvSub n len i) a
+    (\ (j:bitvector n) (pf2:is_bvult n j (bvSub n len i)) ->
+       atBVVec n len a v (bvAdd n i j) (bvult_sub_add_bvult n j i len pf pf2));
+
+-- Combined drop and take
+sliceBVVec : (n : Nat) -> (len : bitvector n) -> (a : sort 0) ->
+             (start' len' : bitvector n) -> is_bvule n start' len ->
+             is_bvule n len' (bvSub n len start') ->
+             BVVec n len a -> BVVec n len' a;
+sliceBVVec n len a start' len' pf1 pf2 v =
+  takeBVVec n (bvSub n len start') a len' pf2
+    (dropBVVec n len a start' pf1 v);
+
+-- Update a slice of a BVVec
+updSliceBVVec : (n : Nat) -> (len : bitvector n) -> (a : sort 0) ->
+                BVVec n len a -> (start' len' : bitvector n) ->
+                BVVec n len' a -> BVVec n len a;
+updSliceBVVec n len a v start' len' v_sub =
+  genBVVec n len a
+    (\ (i:bitvector n) (pf:is_bvult n i len) ->
+       ite a (bvule n start' i)
+           (maybe (is_bvult n (bvSub n i start') len') a
+              (atBVVec n len a v i pf)
+              (\ (pf_sub:is_bvult n (bvSub n i start') len') ->
+                 atBVVec n len' a v_sub (bvSub n i start') pf_sub)
+              (bvultWithProof n (bvSub n i start') len'))
+           (atBVVec n len a v i pf));
+
+
+-------------------------------------------------------------------------------
+-- Dependent sums
+
+data Sigma (a : sort 0) (b : a -> sort 0) : sort 0 where {
+    exists : (pa : a) -> b pa -> Sigma a b;
+}
+
+Sigma__rec (a : sort 0) (b : a -> sort 0) (p : Sigma a b -> sort 0)
+           (f1 : (pa : a) -> (pb : b pa) -> p (exists a b pa pb)) (u : Sigma a b) : p u
+  = Sigma#rec a b p f1 u;
+
+Sigma_proj1 : (a : sort 0) -> (b : a -> sort 0) -> Sigma a b -> a;
+Sigma_proj1 a b =
+  Sigma__rec a b (\ (_:Sigma a b) -> a) (\ (pa:a) (_: b pa) -> pa);
+
+Sigma_proj2 : (a : sort 0) -> (b : a -> sort 0) -> (p : Sigma a b) ->
+              b (Sigma_proj1 a b p);
+Sigma_proj2 a b =
+  Sigma__rec a b (\ (p:Sigma a b) -> b (Sigma_proj1 a b p))
+             (\ (pa:a) (pb: b pa) -> pb);
+
+
+--------------------------------------------------------------------------------
+-- Lists
+
+data List (a:sort 0) : sort 0 where {
+  Nil : List a;
+  Cons : a -> List a -> List a;
+}
```

Comment:
A `List` datatype seems like it would make sense in the `master` saw-core prelude. We should make a ticket asking for it.

Reply:
Agreed!